libsolidity/analysis/ControlFlowRevertPruner.cpp

```diff
@@ -113,6 +113,8 @@ void ControlFlowRevertPruner::findRevertStates()
 				if (
 					item.contract == nullptr ||
 					nextItem.contract == nullptr ||
+					item.contract->isLibrary() ||
+					nextItem.contract->isLibrary() ||
```

Comment:
The ``nullptr`` cases above are for free functions, I guess, but it's the same for library functions, isn't it?

Reply:
If this fix is correct, then there should actually be simpler cases to trigger this, without importing anything, etc.
But yeah, not sure, I only marginally looked through it to find this.

Reply:
I think the fix is correct. I spent some time digging into it and testing various scenarios and I think I know what's happening here.

Here's a single-file repro:
```solidity
contract Bar {
    function example() public {
        foo();
        return;
    }

    function foo() internal {
        Foo.nop();
    }
}

contract Y is Bar {}

library Foo {
    function nop() internal {}
}
```

This confirms your suspicion - this has nothing to do with imports. it's all about the order in which functions are processed. All the functions in the snippet above depend on `Foo.nop()`, which gets processed last so they need to be requeued after we know whether it reverts. `Foo` is a library so it does not happen. Your change fixes that.

The warning about unreachable code stops popping up when we modify `example()` to call `Foo.nop()` directly because for `Foo.nop()` the revert status is known. It's only for the other two that it stays at `RevertState::Unknown`. I think that `ControlFlowRevertPruner::findRevertStates()` should really have an assert at the end to ensure that there's no `m_functions` entry left with `RevertState::Unknown`.

I poked around to see if there might be cases other than libraries where things go wrong, but I think it's all fine now. 

The condition that contracts must be the same looked a bit suspicious to me but turns out it's there to handle virtual calls properly. For every contract `m_functions` contains pairs not only between the contract and its own functions but also with all inherited ones. For example in the test case from this PR we have the following pairs:
- (`Foo`, `nop()`),
- (`Bar`, `foo()`),
- (`Bar`, `example()`),
- (`Test`, `foo()`),
- (`Test`, `example()`)`.

For overridden virtuals you get extra pairs, e.g. if `Test` was overriding `foo()`, there would two (`Test`, `foo()`) pairs, one with `foo()` from `Bar` and one with the new one from `Test`. So since every contract has pairs with all the functions it might call, this part of the condition is here to filter out the superfluous cases like `Bar.example() -> Test.foo()` and `Test.example() -> Bar.foo()`.